pkg/auth/requests/token_review.go

```diff
@@ -0,0 +1,78 @@
+package requests
+
+import (
+	"net/http"
+
+	"github.com/rancher/rancher/pkg/auth/tokens"
+	"github.com/rancher/steve/pkg/auth"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/authentication/user"
+	authv1 "k8s.io/client-go/kubernetes/typed/authentication/v1"
+)
+
+var (
+	allowedPaths = map[string]bool{
+		// This is to enable prometheus to use
+		// mounted service account tokens to
+		// fetch metrics.
+		"/metrics": true,
+	}
+)
+
+type TokenReviewAuth struct {
+	AuthClient           authv1.AuthenticationV1Interface
+	InitialAuthenticator auth.Authenticator
+}
+
+func NewTokenReviewAuth(authenticator auth.Authenticator, authClient authv1.AuthenticationV1Interface) auth.Authenticator {
+	return &TokenReviewAuth{
+		AuthClient:           authClient,
+		InitialAuthenticator: authenticator,
+	}
+}
+
+// Authenticate attempts to authenticate using given function. If authentication
+// fails AND the endpoint is in allowedPaths, it will attempt to perform a token
+// review to authenticate token and extract user info.
+func (t *TokenReviewAuth) Authenticate(req *http.Request) (user.Info, bool, error) {
+	info, hasAuth, err := t.InitialAuthenticator.Authenticate(req)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if info.GetName() != "system:cattle:error" {
+		// auth has succeeded
+		return info, hasAuth, nil
+	}
+
+	if !allowedPaths[req.URL.Path] {
```

Comment:
Instead of checking paths here, why not just add this to only the /metrics route? https://github.com/rancher/rancher/pull/36220/files#diff-80627fe013488f5294c7e3c058e6231d1f4d622b3e7dd817bbf9db3c88af0cbbR105

Reply:
Changed to do just this.